sql*: use pgerror in more places

knz opened the pull request:
First commit from #35867.

... instead of errors.New, errors.Errorf etc.

(Pursuant to #35847).

Release note: None

knz committed:
sql*: use pgerror in more places

... instead of errors.New, errors.Errorf etc.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35845)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35845#-:-LaFnSRl8ejU6BPAbc2Y:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bobvawter and @RaduBerinde)



<!-- Sent from Reviewable.io -->

knz committed:
sql*: distinguish the XX "important" pg errors

Prior to this patch, we were transforming any non-pgerror.Error error
into pg code XX000 ("internal error") when sending it back to the
client.

This is problematic for two reasons:

- the recent changes to make internal errors reportable to telemetry go
  hand-in-hand with a general recommendation to user to inform us
  when they encounter an internal error or code "XX000".

  We do not want *as much* users to inform us when they encounter
  innocuous errors that happen to miss a pg code.

- we want a linter in logic tests to assert that *expected* errors
  that come up during tests are not wrapped in-between the point they
  were generated to the point they are sent to the client by an
  internal error code (as such can happen when an error is encountered
  on a path where no error was expected). In those cases, the expected
  error message will be present (enclosed) in the internal error, and
  thus match the logic test expected error string, although it's not
  really the error that was expected any more.

  To properly identify these internal errors we'd like to rely on the
  error code for internal errors. This can work if all unqualified
  errors default to the same code.

This patch fixes the situation by introducing the new code
XXUUU ("unknown error") for all unqualified errors, and adding the
aforementioned logic test error check.

Release note (sql change): regular SQL errors that indicate erroneous
SQL and for which CockroachDB does not yet populate a well-defined
PostgreSQL error code will now be reported with code XXUUU instead of
code XX000.

knz commented:
bors r+
